subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/SelectorState.java

```diff
@@ -106,14 +109,23 @@ public ComponentState resolveModuleRevisionId() {
 
         selected = resolveState.getRevision(idResolveResult.getModuleVersionId());
         selected.selectedBy(this);
-        selected.addCause(idResolveResult.getSelectionDescription());
+        selected.addCause(doGetDescription());
         targetModule = selected.getModule();
         targetModule.addSelector(this);
         versionConstraint = idResolveResult.getResolvedVersionConstraint();
 
         return selected;
     }
 
+    private ComponentSelectionDescriptorInternal doGetDescription() {
+        ComponentSelectionDescriptorInternal selectionDescription = idResolveResult.getSelectionDescription();
+        if (selectionDescription.getCause() ==  ComponentSelectionCause.REQUESTED && dependencyMetadata.isPending()) {
+            selectionDescription = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.CONSTRAINT,
+                selectionDescription.hasCustomDescription() ? selectionDescription.getDescription() : ComponentSelectionCause.CONSTRAINT.getDefaultReason());
+        }
```

Comment:
Why can't we add 2 different causes here? The component is _both_ `REQUESTED` and `CONSTRAINT`.

Reply:
I guess you're right, I'll give it a try.